tests/net/icmpv4/src/main.c

```diff
@@ -106,236 +106,405 @@ static const unsigned char icmpv4_echo_req_opt_bad[] = {
+#define TEST_ICMPV4_ECHO_REQ_OPTS 2
+
+static u8_t current = TEST_ICMPV4_UNKNOWN;
+static struct in_addr my_addr  = { { { 192, 0, 2, 1 } } };
+static struct net_if *iface;
+
 static enum net_verdict handle_reply_msg(struct net_pkt *pkt,
 					 struct net_ipv4_hdr *ip_hdr,
 					 struct net_icmp_hdr *icmp_hdr)
 {
-	enum net_verdict ret;
-
-	handler_called++;
-
 	if (net_pkt_get_len(pkt) != sizeof(icmpv4_echo_rep)) {
-		goto fail;
+		return NET_DROP;
 	}
 
-	handler_status = 0;
-	ret = NET_OK;
+	net_pkt_unref(pkt);
+	return NET_OK;
+}
 
-	return ret;
+static struct net_icmpv4_handler echo_rep_handler = {
+	.type = NET_ICMPV4_ECHO_REPLY,
+	.code = 0,
+	.handler = handle_reply_msg,
+};
 
-fail:
-	handler_status = -EINVAL;
-	ret = NET_DROP;
+struct net_icmpv4_context {
+	u8_t mac_addr[sizeof(struct net_eth_addr)];
+	struct net_linkaddr ll_addr;
+};
 
-	return ret;
+static int net_icmpv4_dev_init(struct device *dev)
+{
+	struct net_icmpv4_context *net_icmpv4_context = dev->driver_data;
+
+	net_icmpv4_context = net_icmpv4_context;
+
+	return 0;
 }
 
-static enum net_verdict handle_request_msg(struct net_pkt *pkt,
-					   struct net_ipv4_hdr *ip_hdr,
-					   struct net_icmp_hdr *icmp_hdr)
+static u8_t *net_icmpv4_get_mac(struct device *dev)
 {
-	enum net_verdict ret;
+	struct net_icmpv4_context *context = dev->driver_data;
+
+	if (context->mac_addr[2] == 0x00) {
+		/* 00-00-5E-00-53-xx Documentation RFC 7042 */
+		context->mac_addr[0] = 0x00;
+		context->mac_addr[1] = 0x00;
+		context->mac_addr[2] = 0x5E;
+		context->mac_addr[3] = 0x00;
+		context->mac_addr[4] = 0x53;
+		context->mac_addr[5] = 0x01;
+	}
 
-	handler_called++;
+	return context->mac_addr;
+}
 
-	if (net_pkt_get_len(pkt) != sizeof(icmpv4_echo_req)) {
-		goto fail;
+static void net_icmpv4_iface_init(struct net_if *iface)
+{
+	u8_t *mac = net_icmpv4_get_mac(net_if_get_device(iface));
+
+	net_if_set_link_addr(iface, mac, 6, NET_LINK_ETHERNET);
+}
+
+static int verify_echo_reply(struct net_pkt *pkt)
+{
+	struct net_icmp_hdr icmp_hdr;
+	u8_t buf[60];
+	int ret;
+	u8_t payload_len;
+
+	net_pkt_set_overwrite(pkt, true);
+	net_pkt_cursor_init(pkt);
+
+	ret = net_pkt_skip(pkt, NET_IPV4H_LEN);
+	if (ret != 0) {
+		zassert_true(false, "echoeply skip failed");
 	}
 
-	handler_status = 0;
-	ret = NET_OK;
+	/* EchoReply Code and Type is 0 */
+	ret = net_pkt_read(pkt, &icmp_hdr, sizeof(struct net_icmp_hdr));
+	if (ret != 0) {
+		zassert_true(false, "echoeply read failed");
+	}
 
-	return ret;
+	if (icmp_hdr.code != 0 || icmp_hdr.type != 0) {
+		zassert_true(false, "echoeply invalid type or code");
+	}
 
-fail:
-	handler_status = -EINVAL;
-	ret = NET_DROP;
+	/* Calculate payload length */
+	payload_len = sizeof(icmpv4_echo_req) -
+		      NET_IPV4H_LEN - NET_ICMPH_LEN;
+	if (payload_len != net_pkt_remaining_data(pkt)) {
+		zassert_true(false, "echoeply invalid payload len");
+	}
+
+	ret = net_pkt_read(pkt, buf, payload_len);
+	if (ret != 0) {
+		zassert_true(false, "echoeply read payload failed");
```

Comment:
same here, actually there were already echoeply strings earlier in the file. Could you search and fix them all

Reply:
sorry, done.